ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -12143,6 +12148,27 @@ samplePredicate, true, new SampleDesc(ts.getNumerator(),
 
     Operator output = putOpInsertMap(op, rwsch);
 
+    if (tab.isMaterializedTable()) {
+      final FileSinkOperator source = ctx.getMaterializedTableSource(tab.getFullTableName());
+      final Statistics stats = source.getStatistics().clone();
+      final List<ColStatistics> sourceColStatsList = stats.getColumnStats();
+      final List<String> colNames = tab.getCols().stream().map(FieldSchema::getName).collect(Collectors.toList());
+      if (sourceColStatsList.size() != colNames.size()) {
+        throw new IllegalStateException(String.format(
+            "The size of col stats must be equal to that of schema. Stats = %s, Schema = %s",
+            sourceColStatsList, colNames));
+      }
+      final List<ColStatistics> colStatsList = new ArrayList<>(sourceColStatsList.size());
+      for (int i = 0; i < sourceColStatsList.size(); i++) {
+        final ColStatistics colStats = sourceColStatsList.get(i).clone();
+        // FileSinkOperator stores column stats with internal names such as "_col1"
+        colStats.setColumnName(colNames.get(i));
+        colStatsList.add(colStats);
+      }
+      stats.setColumnStats(colStatsList);
```

Comment:
What is the reason of cloning `ColStatistics` objects again?
`Statistics.clone()` does it too
https://github.com/apache/hive/blob/5b76949da6fe65364a4e3766680871167131157f/ql/src/java/org/apache/hadoop/hive/ql/plan/Statistics.java#L210-L216

Reply:
Just because `stats` could become temporarily inconsistent on `colStats.setColumnName`. Now, I think it is not a problem and I will remove the additional clone.